### PR TITLE
Add predefined date dimension for the Pivot Grid

### PR DESCRIPTION
**Pull request by skrustev**
Closes #  

### Additional information (check all that apply):
 - [ ] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [ ] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts [hunk 1]
```diff
@@ -57,3 +58,78 @@ export enum PivotDimensionType {
...
+//                 childLevel: {
+//                         member: 'Date',
+//                         fieldName:'Date',
+//                         enabled: true
+//                     }
+//             }
+//     }
+// },
+
+export class IgxPivotDateDimension implements IPivotDimension {
```
Comment (dkamburov) [inline projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:90]: I guess this should be in another file, not in interfaces

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts [hunk 2]
```diff
@@ -57,3 +58,78 @@ export enum PivotDimensionType {
...
+            memberFunction: (rec) => {
+                const recordValue = PivotUtil.extractValueFromDimension(inChildLevel, rec);
+                return recordValue ? (new Date(recordValue)).getFullYear().toString() : rec['Years'];
+            },
+            enabled: true,
+            childLevel: showQuarters ? quarterDimension : monthDimension
+        };
+    }
+
+    public memberFunction() {
```
Comment (dkamburov) [inline projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:132]: Looks like obsolete. Is it?

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts [hunk 3]
```diff
@@ -57,3 +58,78 @@ export enum PivotDimensionType {
...
+//                         fieldName:'Date',
+//                         enabled: true
+//                     }
+//             }
+//     }
+// },
+
+export class IgxPivotDateDimension implements IPivotDimension {
+    public childLevel: IPivotDimension;
+    public memberName = 'AllPeriods';
```
Comment (dkamburov) [inline projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:92]: If we include 'AllPeriods' it needs to be part of of the localized strings

**Review by dkamburov — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts [hunk 4]
```diff
@@ -57,3 +58,78 @@ export enum PivotDimensionType {
     Row,
     Column
 }
+
+// Equals to pretty much this configuration:
```
Comment (ChronosSF) [inline projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:62]: Nice try skipping the proper format for the API docs ;)

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts (hunk 3 above)
Comment (ChronosSF) [reply projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:92]: If it's something to be localized, it should be human-readable like 'All Periods'

**Review by ChronosSF — APPROVED** [review]

**Comment by ChronosSF** [comment]
Agree with @dkamburov 's comments, other than that looks very nice.

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts (hunk 4 above)
Comment (skrustev) [reply projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:62]: Do we have any way of writing descriptions for classes anyway?

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts (hunk 3 above)
Comment (skrustev) [reply projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:92]: This is not the text that needs localizing, its the value that `memberFunction` returns that needs localizing.

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts (hunk 4 above)
Comment (ChronosSF) [reply projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:62]: you can still leave a description on that class using the /** */ notation that will show up when the user tries typing `new IgxPivot...` etc . And add a sample usage if you like. Doesn't have to be super special.

File: projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts (hunk 2 above)
Comment (dkamburov) [reply projects/igniteui-angular/src/lib/grids/pivot-grid/pivot-grid.interface.ts:132]: Yeah, that's right. I got it wrong

**Comment by MayaKirova** [comment]
@skrustev 
I think this record should be merged with the parent since they share a root:
image

Otherwise the child record will not be flattened into the final collection: 
image

It will be missing under All:
image
 
The data should look similar to this: http://jsfiddle.net/b9qg48s5/6/ :
 
image

**Comment by MayaKirova** [comment]
I also noticed that with the following config;

```
rows: [
            new IgxPivotDateDimension(
                {
                    memberName: 'Date',
                    enabled: true
                },
                {
                    months: false
                }
            ),
            {
                memberName: 'City',
                enabled: true
            },
            {
                memberFunction: () => 'All',
                memberName: 'AllProducts',
                enabled: true,
                childLevel: {
                    memberFunction: (data) => data.ProductCategory,
                    memberName: 'ProductCategory',
                    enabled: true
                }
            }
        ],
```
There are more records than there should be:

image

Notice there are 4 records for Plovdiv, but in the original data there's only 1 record Plovdiv:
image
With category Clothing. So I'm not sure where the other 2 records (Accessories and Components) are coming from. They should not be there.

**Comment by MayaKirova** [comment]
The only issue that remain is that with the following config:
```
  rows: [
            new IgxPivotDateDimension(
                {
                    memberName: 'Date',
                    enabled: true
                },
                {
                    months: false
                }
            ),
            {
                memberName: 'City',
                enabled: true
            },
            {
                memberFunction: () => 'All',
                memberName: 'AllProducts',
                enabled: true,
                childLevel: {
                    memberFunction: (data) => data.ProductCategory,
                    memberName: 'ProductCategory',
                    enabled: true
                }
            }
        ],
```
There are more records than there should be under the date hierarhy:

image


The original data has only 2 records for 
```
 {
            ProductCategory: 'Bikes', UnitPrice: 3.56, SellerName: 'Lydia Burson',
            Country: 'Uruguay', City: 'Ciudad de la Costa', Date: '01/06/2011', UnitsSold: 68
        },
...
 {
            ProductCategory: 'Clothing', UnitPrice: 68.33, SellerName: 'Larry Lieb',
            Country: 'Uruguay', City: 'Ciudad de la Costa', Date: '05/12/2011', UnitsSold: 456
        },

```

And none of them are for 2012, so there should be no records for Ciudad de la Costa under 2012.

**Review by MayaKirova — APPROVED** [review]

**Comment by dkamburov** [comment]
@skrustev there are lint errors.

**Review by dkamburov — APPROVED** [review]

**Pull request merged**